pkg/controller/deployment/deployment_controller.go

```diff
@@ -564,6 +567,58 @@ func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extension
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
 }
 
+// proportionalScaling should run only on scaling events and not during the normal rollout process.
+func (dc *DeploymentController) proportionalScaling(deployment extensions.Deployment) error {
+	newRS, oldRSs, err := dc.getAllReplicaSets(deployment)
+	if err != nil {
+		return err
+	}
+	if len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+		// No old replica sets with pods; scale new replica set to the full count.
+		_, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment)
+		return err
+	}
+	allRSs := append(controller.FilterActiveReplicaSets(oldRSs), newRS)
+
+	sort.Sort(controller.ByAvailableReplicas(allRSs))
+
+	maybeMaxSurge := 0
+	if deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType {
+		maybeMaxSurge, err = intstrutil.GetValueFromIntOrPercent(&deployment.Spec.Strategy.RollingUpdate.MaxSurge, deployment.Spec.Replicas)
```

Comment:
maybeMaxSurge won't be correct in the case that newRc.Spec.Replicas has already reached deployment.Spec.Replicas.

We need to do something like `maybeMaxSurge = IntMax(IntMin(maybeMaxSurge, newRc.Spec.Replicas - deployment.Spec.Replicas), 0)`